### ISSUE TITLE
"tried to create Proc object without a block"

### ISSUE THREAD

**Issue by idanab-simplee** [yammer/circuitbox#185]
Hi

I am trying to use circuitbox for the first time. In my basic usage:

```
def circuit
    Circuitbox::CircuitBreaker.new(:my_service, :exceptions => [Timeout::Error])
end

circuit.run do
    Timeout::timeout(3) do
      response = [http call to some service]
    end
end
```

When I run this code I am getting:
```/usr/local/bundle/gems/circuitbox-1.1.1/lib/circuitbox/circuit_breaker.rb:79:in `new': tried to create Proc object without a block (ArgumentError)```

anyone knows what I am doing wrong?

**Comment by matthewshafer** [comment]
What is the ruby version being used when the error happens?

**Comment by idanab-simplee** [comment]
Hi @matthewshafer,
I am running with ruby 3.1.3p185.

**Comment by idanab-simplee** [comment]
@matthewshafer it is possible that the gem is now breaking because of this breaking change in ruby:
https://github.com/aws/aws-sdk-ruby/issues/2058

this is exactly what is executed in 
circuitbox-1.1.1/lib/circuitbox/circuit_breaker.rb:79

**Comment by idanab-simplee** [comment]
I can verify that the problem is solved in version 2.0.0.pre4

**Comment by matthewshafer** [comment]
The 1.x releases are quite old, though they are the latest non pre-release of circuitbox, and were only tested through ruby 2.2. The 2.x releases have had significant internal changes and have been tested on newer releases of ruby (>= 2.6 at the current time) so if possible I would try and use those versions. There are some changes to circuits between 2.0.0.pre4 and main but those should be reflected in the readme.

**Comment by idanab-simplee** [comment]
I believe that we won't be able to use Circuitbox eventually because we cannot use pre-releases (we have some security limitations). 
Thanks a lot for helping anyway

**Comment by ritikesh** [comment]
@matthewshafer is there anything pending from making an official release? Circuitbox has remained in a pre-release state for a long while now and a lot has changed since then. Is there anything major that you're still planning on adding before making an official 2.0 release?

**Comment by TL-David-Bauske** [comment]
@matthewshafer If it's not possible to release 2.0 shortly, would you accept a PR to the 1.x version that fixes this Ruby 3 incompatibility and re-release it? This issue is stopping our Ruby 3 upgrade.

**Comment by ritikesh** [comment]
@TL-David-Bauske - this might help - https://github.com/yammer/circuitbox/issues/137#issuecomment-1443219012

**Comment by matthewshafer** [comment]
> @matthewshafer If it's not possible to release 2.0 shortly, would you accept a PR to the 1.x version that fixes this Ruby 3 incompatibility and re-release it? This issue is stopping our Ruby 3 upgrade.

I would need to make a 1.x branch, backport the actions from main to that branch, and then fix the reported issue and any other ruby 3 incompatibilities there are. This would take away whatever time I could spend wrapping up work on 2.0.0.

I have been spending what time I can allocate to circuitbox, working on circuitbox. I realize this isn't getting done fast enough for a lot of people who use the project. The recent work I have been doing is time window improvements, one was recently merged, and another one that would improve time window performance when using the in memory store. Additionally I've also been working on documentation that is needed before 2.0.0 can be released. I'm planning to release another prerelease with the time window improvements after I merge or decide not to move forward with the changes I have been working on.

**Comment by TL-David-Bauske** [comment]
@matthewshafer Thanks for the reply and all your work on circuitbox, I didn't mean to sound ungrateful.

I see your point and it makes sense. If this becomes a pressing issue for us, I could volunteer to do the work of backporting the fixes to 1.0, with your permission and approval of course.

**Comment by hank-spokeo** [comment]
I created this PR which appears to fix the Ruby 3 compatibility issue. I couldn't get the tests to run and I doubt the 6-year old test suite would work on Ruby 3 without significant modifications. However, our project does have an integration test where we spin up a local `TCPServer` and verify `Circuitbox.instance.circuits['127.0.0.1'].open?` opens and closes as expected, so I'm fairly confident that it works for our use case at least

**Comment by matthewshafer** [comment]
> @matthewshafer Thanks for the reply and all your work on circuitbox, I didn't mean to sound ungrateful.
> 
> I see your point and it makes sense. If this becomes a pressing issue for us, I could volunteer to do the work of backporting the fixes to 1.0, with your permission and approval of course.

@TL-David-Bauske I posted an update over on #137. Hopefully the timeline is useful. If it's possible to try** 2.0.0.pre5 when I release that, it would be great to know if it's working for how you use circuitbox. I've made many changes to the internals of circuitbox to focus on performance and reliability, it's been working really well for us and we use it on some really heavy utilized external resources.

** I realize you may not be able to run it on production, since it's a pre-release but if possible try it against local development / test suite / staging.

@hank-spokeo thanks for creating the PR. I'll keep it open since you've provided the monkey patch someone could use to get 1.x running on ruby 3. Since there's possibly other things I would need to fix to get full ruby 3 support in 1.x I'd rather focus on 2.0.0 with the time I have. Once 2.0.0 is released I'll close your PR, but I do thank you for opening one, it'll probably help others who can't switch to 2.0.0 until I don't release a pre-release.

**Comment by matthewshafer** [comment]
This is fixed in 2.0.0, which is available on rubygems.
